[WebXR] Enable implicit clear for WebXROpaqueFramebuffer

djg opened the pull request:
#### 153d8bbf64370ede7b39c06f77ee8183ec993ee2
<pre>
[WebXR] Enable implicit clear for WebXROpaqueFramebuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=268042">https://bugs.webkit.org/show_bug.cgi?id=268042</a>
<a href="https://rdar.apple.com/117090654">rdar://117090654</a>

Reviewed by Kimmo Kinnunen.

WebGL&apos;s default framebuffer has an implicit clear behavior. The WebXR
specification requires the same behavior for its WebXROpaqueFramebuffer,
therefore WebXR must ensure that it&apos;s buffers are always cleared for each rAF of
the session.

Currently we assume content does not do redundant initial clear, as the spec
says the buffer always starts cleared.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/html/canvas/WebGLUtilities.h:
(WebCore::ScopedClearColorAndMask::ScopedClearColorAndMask):
(WebCore::ScopedClearColorAndMask::~ScopedClearColorAndMask):
(WebCore::ScopedClearDepthAndMask::ScopedClearDepthAndMask):
(WebCore::ScopedClearDepthAndMask::~ScopedClearDepthAndMask):
(WebCore::ScopedClearStencilAndMask::ScopedClearStencilAndMask):
(WebCore::ScopedClearStencilAndMask::~ScopedClearStencilAndMask):

Canonical link: <a href="https://commits.webkit.org/274121@main">https://commits.webkit.org/274121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e8de6d20039d99afdabf979fd4d283465465cc38

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37986 "Failed to checkout and rebase branch from PR 23209") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16894 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/40242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40531 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/33794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/39571 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19593 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/14199 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38559 "Failed to checkout and rebase branch from PR 23209") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/14237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/40242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/12433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/40242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41805 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/34460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/40242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/12967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/10606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/36435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/14532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/40242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13390 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4926 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13984 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLFramebufferBase.h:
```diff
@@ -0,0 +1,59 @@
+class WebGLRenderingContextBase;
+
+// Interface to state related to a set of framebuffer attachments.
+class WebGLFramebufferBase {
```
These are only correctly maintained for WebGLDefaultFramebuffer, WebXROpaqueFramebuffer.
These are not correctly maintained for WebGLFramebuffer.

The correct, minimal interface would be:
```
class WebGLFramebufferBase {
public:
         void clearIfNeeded(...) { }
};

```

However, I don't think this is needed for WebXR.

kkinnunen-apple reviewed (commented):
I don't think it's correct.
WebXR framebuffer must be cleared before session rAF, regardless of whether the client draws anything or not.

I think your initial problem to solve is "WebXR does not clear the session framebuffer before rAF".
The solution should be "clear the session framebuffer before rAF".

I think we could focus on optimizing clears if they are problems.
The WebGL clear optimizaiton was a more pressing matter when the preserveDrawingBuffer was default and/or  supported. WebXR doesn't have these issues.

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -252,11 +252,109 @@ class ScopedWebGLRestoreTexture {
+    }
+
+private:
+    WebGLRenderingContextBase& m_context;
```
could store the m_context as ptr instead of having m_enabled

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLUtilities.h:
```diff
@@ -252,11 +252,109 @@ class ScopedWebGLRestoreTexture {
+    }
+
+private:
+    WebGLRenderingContextBase& m_context;
```
could store the context as ptr instead of having m_enabled

djg committed:
[WebXR] Enable implicit clear for WebXROpaqueFramebuffer
https://bugs.webkit.org/show_bug.cgi?id=268042
rdar://117090654

Reviewed by Kimmo Kinnunen.

WebGL's default framebuffer has an implicit clear behavior. The WebXR
specification requires the same behavior for its WebXROpaqueFramebuffer,
therefore WebXR must ensure that it's buffers are always cleared for each rAF of
the session.

Currently we assume content does not do redundant initial clear, as the spec
says the buffer always starts cleared.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/html/canvas/WebGLUtilities.h:
(WebCore::ScopedClearColorAndMask::ScopedClearColorAndMask):
(WebCore::ScopedClearColorAndMask::~ScopedClearColorAndMask):
(WebCore::ScopedClearDepthAndMask::ScopedClearDepthAndMask):
(WebCore::ScopedClearDepthAndMask::~ScopedClearDepthAndMask):
(WebCore::ScopedClearStencilAndMask::ScopedClearStencilAndMask):
(WebCore::ScopedClearStencilAndMask::~ScopedClearStencilAndMask):

Canonical link: https://commits.webkit.org/274121@main
